docs/design/coreclr/botr/data_contracts_for_coreclr.md

```diff
@@ -0,0 +1,430 @@
+# Data Contracts for CoreCLR, and the data contract stream
+
+Data contracts represent the manner in which a tool which is not the runtime can reliably understand and observe the behavior of the runtime. Contracts are defined by their documentation, and the runtime describes what contracts are applicable to understanding that runtime.
+
+## Data Contract Stream
+The physical layout of this stream is not defined in this document, but its practical effects are.
```

Comment:
When we design the physical layout, it would be good to design them in a way that native AOT can statically lay this out into at compile/link time and we don't need to build this at startup.

My specific area of concern is various addresses that are known when building the C/C++ code in Runtime.lib, and various offsets/addresses that are known when building native code in ILC. Ideally we'd design the contracts in a way that it's possible to consume the contracts like this (either through link.exe/ld merging various parts of the contract into a single data blob, or through the contract design that allows various parts to come from non-contiguous blocks of memory, if this is not possible to merge during linking).

AFAIK today the information that comes from ILC is not part of the contract (various offsets of things related to `async`) but SOS reads it from the PDB, making assumptions about the name mangling the compiler is doing. I assume those are moving to contracts.

Reply:
Yep, something like that. But with specific attention to the detail that some things are only known in the obj files that comprise the runtime (e.g. heap ranges), but other things are known in obj files that come from ILC (e.g. offset of various fields on Task or async state machines, or whatever managed types we care about). Whether we could design the physical layout in a way that the data can come from two object files that are linked together with a platform linker and be usable as-is (ideally without any fixing up during startup).